File: examples/modulith-clinical-tool-jpa/src/test/java/io/zenwave360/example/clinicaltool/modules/clinical/core/implementation/PatientsServiceTest.java
package io.zenwave360.example.clinicaltool.modules.clinical.core.implementation;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.*;

import io.zenwave360.example.clinicaltool.modules.clinical.config.*;
import io.zenwave360.example.clinicaltool.modules.clinical.core.domain.*;
import io.zenwave360.example.clinicaltool.modules.clinical.core.implementation.mappers.*;
import io.zenwave360.example.clinicaltool.modules.clinical.core.inbound.*;
import io.zenwave360.example.clinicaltool.modules.clinical.core.inbound.dtos.*;
import io.zenwave360.example.clinicaltool.modules.clinical.core.outbound.jpa.*;
import io.zenwave360.example.clinicaltool.modules.clinical.infrastructure.jpa.inmemory.*;
import java.time.*;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Acceptance Test for PatientsService.
 */
class PatientsServiceTest {

    private final Logger log = LoggerFactory.getLogger(getClass());

    ServicesInMemoryConfig context = new ServicesInMemoryConfig();
    PatientsServiceImpl patientsService = context.patientsService();

    PatientRepositoryInMemory patientRepository = context.patientRepository();

    @BeforeEach
    void setUp() {
        context.reloadTestData();
    }

    @Test
    void loadPatientTest() {
    }

    @Test
    void partialPatientUpdateTest() {
    }

    @Test
    void createPatientTest() {
        Patient input = new Patient();
        input.setUserId(1L);
        input.setHospitalId(1L);
        input.setProfilePictureId(1L);
        input.setPhoneNumber("1234567890");
        input.setHisNumber("HIS123456");
        input.setEmail("patient@example.com");

        GeneralInfo generalInfo = new GeneralInfo();
        generalInfo.setName("John");
        generalInfo.setSurname("Doe");
        generalInfo.setSurname2("Smith");
        generalInfo.setIdentityDocumentType(IdentityDocumentType.DNI);
        generalInfo.setIdentityDocumentNumber("12345678A");
        generalInfo.setBirthDate(LocalDate.of(1990, 1, 1));
        generalInfo.setGender(GenderType.MALE);
        generalInfo.setLang("en");
        input.setGeneralInfo(generalInfo);

        HealthInsuranceInfo healthInsuranceInfo = new HealthInsuranceInfo();
        healthInsuranceInfo.setInsuranceCompanyId("INS123");
        healthInsuranceInfo.setInsuranceCardNumber("CARD123456");
        input.setHealthInsuranceInfo(healthInsuranceInfo);

        input.setDocumentIds(List.of(1L, 2L, 3L));
        var patient = patientsService.createPatient(input);
        assertNotNull(patient.getId());
        assertTrue(patientRepository.containsEntity(patient));
    }

    @Test
    void updatePatientTest() {
        Long id = 1L;
        Patient input = new Patient();
        input.setUserId(1L);
        input.setHospitalId(1L);
        input.setProfilePictureId(1L);
        input.setPhoneNumber("0987654321");
        input.setHisNumber("HIS654321");
        input.setEmail("updated.patient@example.com");

        GeneralInfo generalInfo = new GeneralInfo();
        generalInfo.setName("Jane");
        generalInfo.setSurname("Doe");
        generalInfo.setSurname2("Johnson");
        generalInfo.setIdentityDocumentType(IdentityDocumentType.PASSPORT);
        generalInfo.setIdentityDocumentNumber("P87654321");
        generalInfo.setBirthDate(LocalDate.of(1985, 5, 15));
        generalInfo.setGender(GenderType.FEMALE);
        generalInfo.setLang("es");
        input.setGeneralInfo(generalInfo);

        HealthInsuranceInfo healthInsuranceInfo = new HealthInsuranceInfo();
        healthInsuranceInfo.setInsuranceCompanyId("INS456");
        healthInsuranceInfo.setInsuranceCardNumber("CARD654321");
        input.setHealthInsuranceInfo(healthInsuranceInfo);

        input.setDocumentIds(List.of(4L, 5L));
        var patient = patientsService.updatePatient(id, input);
        assertTrue(patient.isPresent());
        assertTrue(patientRepository.containsEntity(patient.get()));
    }

    @Test
    void getPatientTest() {
        Long id = 1L;
        var patient = patientsService.getPatient(id);
        assertTrue(patient.isPresent());
    }

    @Test
    void getPatientProfileByIdTest() {
    }

    @Test
    void requestOptOutTest() {
    }

    @Test
    void associateDocumentWithPatientTest() {
    }
}

File: examples/modulith-klinical-tool-jpa/src/test/kotlin/io/zenwave360/example/clinicaltool/modules/clinical/core/implementation/PatientsServiceTest.kt
package io.zenwave360.example.clinicaltool.modules.clinical.core.implementation

import io.zenwave360.example.clinicaltool.modules.clinical.config.*
import io.zenwave360.example.clinicaltool.modules.clinical.core.domain.*
import io.zenwave360.example.clinicaltool.modules.clinical.core.implementation.mappers.*
import io.zenwave360.example.clinicaltool.modules.clinical.core.inbound.*
import io.zenwave360.example.clinicaltool.modules.clinical.core.inbound.dtos.*
import io.zenwave360.example.clinicaltool.modules.clinical.core.outbound.jpa.*
import io.zenwave360.example.clinicaltool.modules.clinical.infrastructure.jpa.inmemory.*
import java.time.*
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/** Acceptance Test for PatientsService. */
class PatientsServiceTest {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    val context = ServicesInMemoryConfig()
    val patientsService: PatientsService = context.patientsService()

    val patientRepository: PatientRepositoryInMemory = context.patientRepository()

    @BeforeEach
    fun setUp() {
        context.reloadTestData()
    }

    @Test
    fun loadPatientTest() { // TODO: implement this test
    }

    @Test
    fun createPatientTest() {
        val input: Patient = Patient()
        input.userId = 1L
        input.hospitalId = 1L
        input.profilePictureId = 1L
        input.phoneNumber = "1234567890"
        input.hisNumber = "HIS123456"
        input.email = "patient@example.com"

        val generalInfo = GeneralInfo()
        generalInfo.name = "John"
        generalInfo.surname = "Doe"
        generalInfo.surname2 = "Smith"
        generalInfo.identityDocumentType = IdentityDocumentType.DNI
        generalInfo.identityDocumentNumber = "12345678A"
        generalInfo.birthDate = LocalDate.of(1990, 1, 1)
        generalInfo.gender = GenderType.MALE
        generalInfo.lang = "en"
        input.generalInfo = generalInfo

        val healthInsuranceInfo = HealthInsuranceInfo()
        healthInsuranceInfo.insuranceCompanyId = "INS123"
        healthInsuranceInfo.insuranceCardNumber = "CARD123456"
        input.healthInsuranceInfo = healthInsuranceInfo

        input.documentIds = mutableListOf(1L, 2L, 3L)
        val patient = patientsService.createPatient(input)
        assertNotNull(patient.id)
        assertTrue(patientRepository.containsEntity(patient))
    }

    @Test
    fun partialPatientUpdateTest() { // TODO: implement this test
    }

    @Test
    fun updatePatientTest() {
        val id: Long = 1L
        val input: Patient = Patient()
        input.userId = 1L
        input.hospitalId = 1L
        input.profilePictureId = 1L
        input.phoneNumber = "0987654321"
        input.hisNumber = "HIS654321"
        input.email = "updated.patient@example.com"

        val generalInfo = GeneralInfo()
        generalInfo.name = "Jane"
        generalInfo.surname = "Doe"
        generalInfo.surname2 = "Johnson"
        generalInfo.identityDocumentType = IdentityDocumentType.PASSPORT
        generalInfo.identityDocumentNumber = "P87654321"
        generalInfo.birthDate = LocalDate.of(1985, 5, 15)
        generalInfo.gender = GenderType.FEMALE
        generalInfo.lang = "es"
        input.generalInfo = generalInfo

        val healthInsuranceInfo = HealthInsuranceInfo()
        healthInsuranceInfo.insuranceCompanyId = "INS456"
        healthInsuranceInfo.insuranceCardNumber = "CARD654321"
        input.healthInsuranceInfo = healthInsuranceInfo

        input.documentIds = mutableListOf(4L, 5L)
        val patient = patientsService.updatePatient(id, input)
        assertNotNull(patient)
        assertTrue(patientRepository.containsEntity(patient!!))
    }

    @Test
    fun getPatientTest() {
        val id: Long = 1L
        val patient = patientsService.getPatient(id)
        assertNotNull(patient)
    }

    @Test
    fun getPatientProfileByIdTest() { // TODO: implement this test
    }

    @Test
    fun associateDocumentWithPatientTest() { // TODO: implement this test
    }
}
